eva-accession-import-automation/genome_downloader.py

```diff
@@ -0,0 +1,170 @@
+                           dtype=str, sep='\t')
+
+
+def download_assembly_report(assembly_accession, directory_path):
+    assembly_report_url = get_assembly_report_url(assembly_accession)
+    assembly_report_path = directory_path + os.path.basename(assembly_report_url)
+    os.makedirs(os.path.dirname(assembly_report_path), exist_ok=True)
+    urllib.request.urlretrieve(assembly_report_url, assembly_report_path)
+    urllib.request.urlcleanup()
+    return assembly_report_path
+
+
+def get_sequence_from_ncbi(accession, fasta_path, eutils_api_key, written_contigs):
+    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=' + accession + \
+           '&rettype=fasta&retmode=text&api_key=' + eutils_api_key + '&tool=eva&email=eva-dev@ebi.ac.uk'
+    logger.info('Downloading ' + accession)
+    with urllib.request.urlopen(url) as response:
+        data = get_and_process_sequence(response)
+        if is_sequence_valid(data):
+            logger.info('FASTA sequence not available for' + accession)
+        else:
+            contatenate_sequence_to_fasta(fasta_path, data)
+            logger.info(accession + " downloaded and added to FASTA file")
+            written_contigs.append(accession)
+
+
+def get_and_process_sequence(response):
+    retrieved_data = response.read().decode('utf-8')
+    # Delete accession prefixes
+    processed_data = re.sub(r'\s*ENA\|.*\|', '', retrieved_data)
+    return processed_data
+
+
+def is_sequence_valid(data):
+    # If a file has more than one line, then it is concatenated into the full assembly FASTA file
+    # (empty sequences can't be indexed)
+    is_sequence_empty = re.match(r'^>.*\n[ACGTN]+', data) is None
+    return is_sequence_empty
+
+
+def contatenate_sequence_to_fasta(fasta_path, data):
+    with open(fasta_path, 'a+') as file:
+        file.write(str(data))
+
+
+def build_fasta_from_assembly_report(assembly_accession, assembly_report_dataframe, eutils_api_key, directory_path, clear):
+    fasta_path = directory_path + assembly_accession + '.fa'
+    if clear:
+        clear_file_content(fasta_path)
+
+    written_contigs = get_written_contigs(fasta_path)
+    for index, row in assembly_report_dataframe.iterrows():
+        genbank_accession = row['GenBank-Accn']
+        refseq_accession = row['RefSeq-Accn']
+        relationship = row['Relationship']
+        accession = genbank_accession
+
+        if relationship != '=' and genbank_accession == 'na':
+            accession = refseq_accession
+
+        if written_contigs is not None and (genbank_accession in written_contigs or refseq_accession in written_contigs):
+            logger.info('Accession ' + accession + ' already in the FASTA file, don\'t need to be downloaded')
+            continue
+
+        get_sequence_from_ncbi(accession, fasta_path, eutils_api_key, written_contigs)
+
+
+def get_written_contigs(fasta_path):
+    try:
+        with open(fasta_path, 'r') as file:
+            data = file.read()
```

Comment:
Based on how big the sequence is for a given contig, this could occupy a lot of memory and sometimes can even crash. Strongly recommend a line-by-line scan here.....